packages/eth-json-rpc-provider/src/safe-event-emitter-provider.test.ts

```diff
@@ -30,6 +58,92 @@ describe('SafeEventEmitterProvider', () => {
     });
   });
 
+  it('returns the correct block number with @metamask/eth-query', async () => {
+    const provider = new SafeEventEmitterProvider({
+      engine: createMockEngine('eth_blockNumber', 42),
+    });
+    const ethQuery = new EthQuery(provider);
+
+    ethQuery.sendAsync({ method: 'eth_blockNumber' }, (_error, response) => {
+      expect(response).toBe(42);
+    });
+  });
+
+  it('returns the correct block number with @metamask/ethjs-query', async () => {
+    const provider = new SafeEventEmitterProvider({
+      engine: createMockEngine('eth_blockNumber', 42),
+    });
+    const ethJsQuery = new EthJsQuery(provider);
+
+    const response = await ethJsQuery.blockNumber();
+
+    expect(response.toNumber()).toBe(42);
+  });
+
+  it('returns the correct block number with Web3Provider', async () => {
+    const provider = new SafeEventEmitterProvider({
+      engine: createMockEngine('eth_blockNumber', 42),
+    });
+    const web3Provider = new Web3Provider(provider);
+
+    const response = await web3Provider.send('eth_blockNumber', []);
+
+    expect(response).toBe(42);
+  });
+
+  it('returns the correct block number with BrowserProvider', async () => {
+    const provider = new SafeEventEmitterProvider({
+      engine: createMockEngine('eth_blockNumber', 42),
+    });
+    const browserProvider = new BrowserProvider(provider);
+
+    const response = await browserProvider.send('eth_blockNumber', []);
+
+    expect(response).toBe(42);
+  });
+
+  describe('request', () => {
+    it('handles a successful request', async () => {
+      const engine = new JsonRpcEngine();
+      engine.push((_req, res, _next, end) => {
+        res.result = 42;
+        end();
+      });
+      const provider = new SafeEventEmitterProvider({ engine });
+      const exampleRequest = {
+        id: 1,
+        jsonrpc: '2.0' as const,
+        method: 'test',
+      };
+
+      const response = await provider.request(exampleRequest);
+
+      expect(response).toBe(42);
+    });
+
+    it('handles a failed request', async () => {
```

Comment:
Thinking about this some more, I think we need two tests here: one that confirms what happens when an unknown error (non-JSON-RPC error) is thrown in middleware, and one that confirms what happens when the middleware is ended with a JSON-RPC error.

If the request fails with a non-JSON-RPC error, then we want to make sure that the error thrown by `request`:

- has a `message` property of "Internal JSON-RPC error." (this is okay, because this is what `JsonRpcEngine` does)
- has a `code` property whose value is the same as an internal JSON-RPC error, i.e. -32603
- has a `data` property, and the value of `data.cause` is the original error object thrown in the middleware
- has a `stack` property which is not overwritten (i.e. it should include the line that threw the error in the middleware)

If the request fails with a JSON-RPC error, then we want to make sure that the error thrown by `request`:

- has a `message` property that matches the `message` of the JSON-RPC error
- has a `code` property that matches the `code` of the JSON-RPC error
- has a `data` property that matches the `data` of the JSON-RPC error